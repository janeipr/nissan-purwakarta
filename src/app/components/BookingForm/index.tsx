"use client";

import { useState } from "react";
import { IoCloseOutline } from "react-icons/io5";
import CONFIG from "@/config";

interface BookingFormProps {
  isOpen: boolean;
  onClose: () => void;
}

export default function BookingForm({ isOpen, onClose }: BookingFormProps) {
  const [formData, setFormData] = useState({
    name: "",
    address: "",
    phone: "",
    date: "",
    time: "",
    vehicleModel: "",
    serviceType: "",
    notes: "",
  });

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    
    const message = `Hai! Saya ingin reservasi service di Nissan Purwakarta.
Berikut data diri Saya:
Nama: ${formData.name}
Alamat: ${formData.address}
No. Telepon: ${formData.phone}
Hari dan tanggal reservasi: ${formData.date}
Jam reservasi: ${formData.time}
Model Kendaraan: ${formData.vehicleModel}
Jenis Service: ${formData.serviceType}
Catatan Tambahan: ${formData.notes}`;

    window.open(
      `https://wa.me/${CONFIG.WA_NUMBER}?text=${encodeURIComponent(message)}`
    );
    onClose();
  };

  const isWeekday = (date: string) => {
    const day = new Date(date).getDay();
    return day !== 0 && day !== 6;
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>) => {
    const { name, value } = e.target;
    
    if (name === 'date') {
      // Only update if it's a weekday
      if (isWeekday(value)) {
        setFormData((prev) => ({
          ...prev,
          [name]: value,
        }));
      }
      return;
    }

    setFormData((prev) => ({
      ...prev,
      [name]: value,
    }));
  };

  // Get next available weekday
  const getNextWeekday = () => {
    const today = new Date();
    const day = today.getDay();
    const diff = day === 6 ? 2 : day === 0 ? 1 : 0; // If Saturday +2, if Sunday +1, else +0
    today.setDate(today.getDate() + diff);
    return today.toISOString().split('T')[0];
  };

  if (!isOpen) return null;

  return (
    <div className="fixed inset-0 z-50 flex items-center justify-center">
      <div className="absolute inset-0 bg-black bg-opacity-50" onClick={onClose} />
      <div className="relative bg-white rounded-[30px] p-8 w-full max-w-2xl max-h-[90vh] overflow-y-auto">
        <button
          onClick={onClose}
          className="absolute right-6 top-6 text-gray-500 hover:text-gray-700"
        >
          <IoCloseOutline size={24} />
        </button>
        
        <h2 className="text-2xl font-bold mb-6">Booking Service Appointment</h2>
        
        <form onSubmit={handleSubmit} className="space-y-6">
          <div className="grid grid-cols-1 md:grid-cols-2 gap-6">
            <div>
              <label className="block text-sm font-medium text-gray-700 mb-2">
                Nama Lengkap*
              </label>
              <input
                type="text"
                name="name"
                required
                value={formData.name}
                onChange={handleChange}
                className="w-full px-4 py-2 border border-gray-300 rounded-lg focus:ring-2 focus:ring-portto-light-gold focus:border-transparent"
              />
            </div>
            
            <div>
              <label className="block text-sm font-medium text-gray-700 mb-2">
                No. Telepon*
              </label>
              <input
                type="tel"
                name="phone"
                required
                value={formData.phone}
                onChange={handleChange}
                className="w-full px-4 py-2 border border-gray-300 rounded-lg focus:ring-2 focus:ring-portto-light-gold focus:border-transparent"
              />
            </div>
          </div>

          <div>
            <label className="block text-sm font-medium text-gray-700 mb-2">
              Alamat*
            </label>
            <input
              type="text"
              name="address"
              required
              value={formData.address}
              onChange={handleChange}
              className="w-full px-4 py-2 border border-gray-300 rounded-lg focus:ring-2 focus:ring-portto-light-gold focus:border-transparent"
            />
          </div>

          <div className="grid grid-cols-1 md:grid-cols-2 gap-6">
            <div>
              <label className="block text-sm font-medium text-gray-700 mb-2">
                Tanggal Service* (Senin-Jumat)
              </label>
              <input
                type="date"
                name="date"
                required
                value={formData.date}
                onChange={handleChange}
                min={getNextWeekday()}
                onKeyDown={(e) => e.preventDefault()}
                className="w-full px-4 py-2 border border-gray-300 rounded-lg focus:ring-2 focus:ring-portto-light-gold focus:border-transparent [&::-webkit-calendar-picker-indicator]:cursor-pointer"
              />
            </div>
            
            <div>
              <label className="block text-sm font-medium text-gray-700 mb-2">
                Jam Service*
              </label>
              <input
                type="time"
                name="time"
                required
                value={formData.time}
                onChange={handleChange}
                min="08:00"
                max="17:00"
                step="1800"
                className="w-full px-4 py-2 border border-gray-300 rounded-lg focus:ring-2 focus:ring-portto-light-gold focus:border-transparent [&::-webkit-calendar-picker-indicator]:appearance-none [&::-webkit-time-picker-indicator]:appearance-none"
              />
            </div>
          </div>

          <div className="grid grid-cols-1 md:grid-cols-2 gap-6">
            <div>
              <label className="block text-sm font-medium text-gray-700 mb-2">
                Model Kendaraan*
              </label>
              <select
                name="vehicleModel"
                required
                value={formData.vehicleModel}
                onChange={handleChange}
                className="w-full px-4 py-2 border border-gray-300 rounded-lg focus:ring-2 focus:ring-portto-light-gold focus:border-transparent"
              >
                <option value="">Pilih Model</option>
                <option value="Nissan Kicks">Nissan Kicks</option>
                <option value="Nissan Livina">Nissan Livina</option>
                <option value="Nissan Magnite">Nissan Magnite</option>
                <option value="Nissan Serena">Nissan Serena</option>
                <option value="Nissan Terra">Nissan Terra</option>
                <option value="Lainnya">Lainnya</option>
              </select>
            </div>
            
            <div>
              <label className="block text-sm font-medium text-gray-700 mb-2">
                Jenis Service*
              </label>
              <select
                name="serviceType"
                required
                value={formData.serviceType}
                onChange={handleChange}
                className="w-full px-4 py-2 border border-gray-300 rounded-lg focus:ring-2 focus:ring-portto-light-gold focus:border-transparent"
              >
                <option value="">Pilih Service</option>
                <option value="Perawatan Berkala">Perawatan Berkala</option>
                <option value="Perbaikan Umum">Perbaikan Umum</option>
                <option value="Perbaikan AC">Perbaikan AC</option>
                <option value="Tune Up">Tune Up</option>
                <option value="Ganti Oli">Ganti Oli</option>
                <option value="Lainnya">Lainnya</option>
              </select>
            </div>
          </div>

          <div>
            <label className="block text-sm font-medium text-gray-700 mb-2">
              Catatan Tambahan
            </label>
            <textarea
              name="notes"
              value={formData.notes}
              onChange={handleChange}
              rows={3}
              className="w-full px-4 py-2 border border-gray-300 rounded-lg focus:ring-2 focus:ring-portto-light-gold focus:border-transparent"
              placeholder="Jelaskan keluhan atau permintaan khusus Anda..."
            />
          </div>

          <div className="flex justify-end gap-4">
            <button
              type="button"
              onClick={onClose}
              className="px-6 py-2 border border-gray-300 rounded-full hover:bg-gray-50"
            >
              Batal
            </button>
            <button
              type="submit"
              className="px-6 py-2 bg-portto-light-gold text-black rounded-full hover:shadow-[0_10px_20px_0_#FFE7C280] font-bold"
            >
              Kirim via WhatsApp
            </button>
          </div>
        </form>
      </div>
    </div>
  );
} 