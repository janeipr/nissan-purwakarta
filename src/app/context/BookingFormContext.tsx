"use client";

import { createContext, useContext, useState, ReactNode } from "react";
import BookingForm from "../components/BookingForm";

interface BookingFormContextType {
  isOpen: boolean;
  openForm: () => void;
  closeForm: () => void;
}

const BookingFormContext = createContext<BookingFormContextType | undefined>(undefined);

export function BookingFormProvider({ children }: { children: ReactNode }) {
  const [isOpen, setIsOpen] = useState(false);

  const openForm = () => setIsOpen(true);
  const closeForm = () => setIsOpen(false);

  return (
    <BookingFormContext.Provider value={{ isOpen, openForm, closeForm }}>
      {children}
      <BookingForm isOpen={isOpen} onClose={closeForm} />
    </BookingFormContext.Provider>
  );
}

export function useBookingForm() {
  const context = useContext(BookingFormContext);
  if (context === undefined) {
    throw new Error("useBookingForm must be used within a BookingFormProvider");
  }
  return context;
} 