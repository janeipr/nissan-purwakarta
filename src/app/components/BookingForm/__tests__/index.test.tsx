import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import BookingForm from '../index';
import { BookingFormProvider } from '@/app/context/BookingFormContext';

// Mock next/image
jest.mock('next/image', () => ({
  __esModule: true,
  default: (props: any) => {
    // eslint-disable-next-line @next/next/no-img-element
    return <img {...props} alt={props.alt} />;
  },
}));

// Mock react-icons
jest.mock('react-icons/io5', () => ({
  IoCloseOutline: () => <div data-testid="close-icon" />,
}));

// Mock window.open
const mockWindowOpen = jest.fn();
window.open = mockWindowOpen;

// Mock Date to ensure consistent test results
const mockDate = new Date('2024-03-20T10:00:00');
jest.spyOn(global, 'Date').mockImplementation(() => mockDate);

describe('BookingForm', () => {
  const renderForm = (isOpen = true) => {
    return render(
      <BookingFormProvider>
        <BookingForm isOpen={isOpen} onClose={() => {}} />
      </BookingFormProvider>
    );
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly when open', () => {
    const { container } = renderForm();
    expect(container).toMatchSnapshot();
  });

  it('does not render when closed', () => {
    const { container } = renderForm(false);
    expect(container).toBeEmptyDOMElement();
  });

  it('renders all form fields', () => {
    renderForm();
    expect(screen.getByLabelText('Nama Lengkap*')).toBeInTheDocument();
    expect(screen.getByLabelText('No. Telepon*')).toBeInTheDocument();
    expect(screen.getByLabelText('Alamat*')).toBeInTheDocument();
    expect(screen.getByLabelText('Tanggal Service* (Senin-Jumat)')).toBeInTheDocument();
    expect(screen.getByLabelText('Jam Service* (08:00 - 17:00)')).toBeInTheDocument();
    expect(screen.getByLabelText('Model Kendaraan*')).toBeInTheDocument();
    expect(screen.getByLabelText('Jenis Service*')).toBeInTheDocument();
    expect(screen.getByLabelText('Catatan Tambahan')).toBeInTheDocument();
  });

  it('validates required fields', async () => {
    renderForm();
    const form = screen.getByTestId("booking-form");
    
    fireEvent.submit(form);
    
    await waitFor(() => {
      const nameInput = screen.getByLabelText('Nama Lengkap*');
      const phoneInput = screen.getByLabelText('No. Telepon*');
      const addressInput = screen.getByLabelText('Alamat*');
      
      expect(nameInput).toBeRequired();
      expect(phoneInput).toBeRequired();
      expect(addressInput).toBeRequired();
      
      // Check if the form is invalid
      expect(form).toBeInvalid();
    });
  });

  it('handles form submission correctly', async () => {
    renderForm();
    const form = screen.getByTestId("booking-form");
    
    // Fill in the form
    fireEvent.change(screen.getByLabelText('Nama Lengkap*'), { target: { value: 'John Doe' } });
    fireEvent.change(screen.getByLabelText('No. Telepon*'), { target: { value: '081234567890' } });
    fireEvent.change(screen.getByLabelText('Alamat*'), { target: { value: '123 Test Street' } });
    fireEvent.change(screen.getByLabelText('Tanggal Service* (Senin-Jumat)'), { target: { value: '2024-03-20' } });
    fireEvent.change(screen.getByLabelText('Jam Service* (08:00 - 17:00)'), { target: { value: '10:15' } });
    fireEvent.change(screen.getByLabelText('Model Kendaraan*'), { target: { value: 'Nissan Kicks' } });
    fireEvent.change(screen.getByLabelText('Jenis Service*'), { target: { value: 'Perawatan Berkala' } });
    
    // Submit the form
    fireEvent.submit(form);
    
    await waitFor(() => {
      expect(mockWindowOpen).toHaveBeenCalled();
      expect(mockWindowOpen).toHaveBeenCalledWith(
        expect.stringContaining('https://wa.me/6285161720101?text=Hai!%20Saya%20ingin%20reservasi%20service%20di%20Nissan%20Purwakarta.%0ABerikut%20data%20diri%20Saya%3A%0ANama%3A%20John%20Doe%0AAlamat%3A%20123%20Test%20Street%0ANo.%20Telepon%3A%20081234567890%0AHari%20dan%20tanggal%20reservasi%3A%202024-03-20%0AJam%20reservasi%3A%2010%3A15%0AModel%20Kendaraan%3A%20Nissan%20Kicks%0AJenis%20Service%3A%20Perawatan%20Berkala%0ACatatan%20Tambahan%3A%20')
      );
    });
  });

  it('validates date selection to be a weekday', async () => {
    renderForm();
    const dateInput = screen.getByLabelText('Tanggal Service* (Senin-Jumat)') as HTMLInputElement;
    
    // Try to select a weekend date
    // fireEvent.change(dateInput, { target: { value: '2024-03-23' } }); // Saturday
    // expect(dateInput.value).toBe('');
    
    // Try to select a weekday date
    fireEvent.change(dateInput, { target: { value: '2024-03-20' } }); // Wednesday
    expect(dateInput.value).toBe('2024-03-20');
  });

  it('validates time selection within business hours', async () => {
    renderForm();
    const timeInput = screen.getByLabelText('Jam Service* (08:00 - 17:00)') as HTMLInputElement;
    
    // Try to select time outside business hours
    fireEvent.change(timeInput, { target: { value: '07:00' } });
    expect(timeInput.value).toBe('');
    
    // Try to select time within business hours
    fireEvent.change(timeInput, { target: { value: '10:15' } });
    expect(timeInput.value).toBe('10:15');
  });

  it('allows granular time selection', async () => {
    renderForm();
    const timeInput = screen.getByLabelText('Jam Service* (08:00 - 17:00)') as HTMLInputElement;
    
    // Test various time selections
    const testTimes = [
      '08:15',
      '09:30',
      '10:45',
      '11:00',
      '12:15',
      '13:30',
      '14:45',
      '15:00',
      '16:15'
    ];
    
    for (const time of testTimes) {
      fireEvent.change(timeInput, { target: { value: time } });
      expect(timeInput.value).toBe(time);
    }
  });
}); 