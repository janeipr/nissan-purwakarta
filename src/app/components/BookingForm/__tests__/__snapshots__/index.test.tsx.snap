// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingForm renders correctly when open 1`] = `
<div>
  <div
    class="fixed inset-0 z-50 flex items-center justify-center"
  >
    <div
      class="absolute inset-0 bg-black bg-opacity-50"
    />
    <div
      class="relative bg-white rounded-[30px] p-8 w-full max-w-2xl max-h-[90vh] overflow-y-auto"
    >
      <button
        class="absolute right-6 top-6 text-gray-500 hover:text-gray-700"
      >
        <div
          data-testid="close-icon"
        />
      </button>
      <h2
        class="text-2xl font-bold mb-6"
      >
        Booking Service Appointment
      </h2>
      <form
        class="space-y-6"
        data-testid="booking-form"
      >
        <div
          class="grid grid-cols-1 md:grid-cols-2 gap-6"
        >
          <div>
            <label
              class="block text-sm font-medium text-gray-700 mb-2"
              for="name"
            >
              Nama Lengkap*
            </label>
            <input
              class="w-full px-4 py-2 border border-gray-300 rounded-lg focus:ring-2 focus:ring-portto-light-gold focus:border-transparent"
              id="name"
              name="name"
              required=""
              type="text"
              value=""
            />
          </div>
          <div>
            <label
              class="block text-sm font-medium text-gray-700 mb-2"
              for="phone"
            >
              No. Telepon*
            </label>
            <input
              class="w-full px-4 py-2 border border-gray-300 rounded-lg focus:ring-2 focus:ring-portto-light-gold focus:border-transparent"
              id="phone"
              name="phone"
              required=""
              type="tel"
              value=""
            />
          </div>
        </div>
        <div>
          <label
            class="block text-sm font-medium text-gray-700 mb-2"
            for="address"
          >
            Alamat*
          </label>
          <input
            class="w-full px-4 py-2 border border-gray-300 rounded-lg focus:ring-2 focus:ring-portto-light-gold focus:border-transparent"
            id="address"
            name="address"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="grid grid-cols-1 md:grid-cols-2 gap-6"
        >
          <div>
            <label
              class="block text-sm font-medium text-gray-700 mb-2"
              for="date"
            >
              Tanggal Service* (Senin-Jumat)
            </label>
            <input
              class="w-full px-4 py-2 border border-gray-300 rounded-lg focus:ring-2 focus:ring-portto-light-gold focus:border-transparent [&::-webkit-calendar-picker-indicator]:cursor-pointer"
              id="date"
              min="2024-03-20"
              name="date"
              required=""
              type="date"
              value=""
            />
          </div>
          <div>
            <label
              class="block text-sm font-medium text-gray-700 mb-2"
              for="time"
            >
              Jam Service* (08:00 - 17:00)
            </label>
            <input
              class="w-full px-4 py-2 border border-gray-300 rounded-lg focus:ring-2 focus:ring-portto-light-gold focus:border-transparent [&::-webkit-calendar-picker-indicator]:appearance-none [&::-webkit-time-picker-indicator]:appearance-none"
              id="time"
              max="17:00"
              min="08:00"
              name="time"
              required=""
              type="time"
              value=""
            />
          </div>
        </div>
        <div
          class="grid grid-cols-1 md:grid-cols-2 gap-6"
        >
          <div>
            <label
              class="block text-sm font-medium text-gray-700 mb-2"
              for="vehicleModel"
            >
              Model Kendaraan*
            </label>
            <select
              class="w-full px-4 py-2 border border-gray-300 rounded-lg focus:ring-2 focus:ring-portto-light-gold focus:border-transparent"
              id="vehicleModel"
              name="vehicleModel"
              required=""
            >
              <option
                value=""
              >
                Pilih Model
              </option>
              <option
                value="Nissan Kicks"
              >
                Nissan Kicks
              </option>
              <option
                value="Nissan Livina"
              >
                Nissan Livina
              </option>
              <option
                value="Nissan Magnite"
              >
                Nissan Magnite
              </option>
              <option
                value="Nissan Serena"
              >
                Nissan Serena
              </option>
              <option
                value="Nissan Terra"
              >
                Nissan Terra
              </option>
              <option
                value="Lainnya"
              >
                Lainnya
              </option>
            </select>
          </div>
          <div>
            <label
              class="block text-sm font-medium text-gray-700 mb-2"
              for="serviceType"
            >
              Jenis Service*
            </label>
            <select
              class="w-full px-4 py-2 border border-gray-300 rounded-lg focus:ring-2 focus:ring-portto-light-gold focus:border-transparent"
              id="serviceType"
              name="serviceType"
              required=""
            >
              <option
                value=""
              >
                Pilih Service
              </option>
              <option
                value="Perawatan Berkala"
              >
                Perawatan Berkala
              </option>
              <option
                value="Perbaikan Umum"
              >
                Perbaikan Umum
              </option>
              <option
                value="Perbaikan AC"
              >
                Perbaikan AC
              </option>
              <option
                value="Tune Up"
              >
                Tune Up
              </option>
              <option
                value="Ganti Oli"
              >
                Ganti Oli
              </option>
              <option
                value="Lainnya"
              >
                Lainnya
              </option>
            </select>
          </div>
        </div>
        <div>
          <label
            class="block text-sm font-medium text-gray-700 mb-2"
            for="notes"
          >
            Catatan Tambahan
          </label>
          <textarea
            class="w-full px-4 py-2 border border-gray-300 rounded-lg focus:ring-2 focus:ring-portto-light-gold focus:border-transparent"
            id="notes"
            name="notes"
            placeholder="Jelaskan keluhan atau permintaan khusus Anda..."
            rows="3"
          />
        </div>
        <div
          class="flex justify-end gap-4"
        >
          <button
            class="px-6 py-2 border border-gray-300 rounded-full hover:bg-gray-50"
            type="button"
          >
            Batal
          </button>
          <button
            class="px-6 py-2 bg-portto-light-gold text-black rounded-full hover:shadow-[0_10px_20px_0_#FFE7C280] font-bold"
            type="submit"
          >
            Kirim via WhatsApp
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
